ystia/yorc
# Bugfix/gh 484 consul transactions

# Pull Request description

## Description of the change

Fixed  Yorc failure to register a deploy task when it attempts to register a task with it associated steps in Consul within a transaction, and the transaction exceeds the max number of operations supported by Consul (64).

### What I did

helper/consulutil/consul_ratelimit_publisher.go:
Added a function ExecuteSplittableTransaction() that will split a transation into n transactions if the number of of operations exceeds 64 (max supported by Consul).
When the transaction has to be splitted, it is possible to specify :
a pre-operation that will be executed as the first operation of the first transaction
a post-operation that will be executed as the last operation of the last transaction


tasks/tasks.go:
Added a function StoreOperations() that will call ExecuteSplittableTransaction() with a pre-operation adding a key under the taskID subtree
and a post-operations removing this key.
Added a function IsStepRegistrationInProgress() that will check if this key is present so that callers known if transactions are being executed to add steps to a given task ID.

tasks/collector/collector.go:
Call tasks.StoreOperations() instead of consul call kv.Txn() to split a transaction if it exceeds the max number of operatiosn supported by Consul.

tasks/workflow/dispatcher.go:

Call tasks.IsStepRegistrationInProgress() to know if all steps have been registered in Consul, if not the execution will not be performed in this round.

tasks/workflow/step.go:
Call tasks.StoreOperations() instead of consul call kv.Txn() to split a transaction if it exceeds the max number of operatiosn supported by Consul.

All other changes are related to tests


### How to verify it

Upload this topoly template in Alien4Cloud catalog:
[bigtopology.zip](https://github.com/ystia/yorc/files/3396732/bigtopology.zip)

Create an application from this toplogy template, and deploy this application on OpenStack.
Check the deployment in successful and created 70 compute instances.
Undeploy the application.

### Description for the changelog

Failure to deploy/undeploy big application: Transaction contains too many operations ([GH-484](https://github.com/ystia/yorc/issues/484))

## Applicable Issues

Fixes #484
